File: node/fluent-cards-tests/test/elements.test.ts
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import {
  AdaptiveCardBuilder,
  ContainerBuilder,
  ColumnSetBuilder,
  FactSetBuilder,
  RichTextBlockBuilder,
  ActionSetBuilder,
  MediaBuilder,
  ImageSetBuilder,
  TableBuilder,
  toJson,
  ContainerStyle,
  VerticalAlignment,
  HorizontalAlignment,
  ImageSize,
  TextColor,
  TextWeight,
} from 'fluent-cards';
import type { Container, Column, FactSet, RichTextBlock, Table, TableRow, TableCell, TableColumnDefinition } from 'fluent-cards';

describe('ContainerBuilder', () => {
  it('builds a Container with items', () => {
    const container = new ContainerBuilder()
      .withId('c1')
      .withStyle(ContainerStyle.Emphasis)
      .withVerticalContentAlignment(VerticalAlignment.Center)
      .withBleed()
      .withMinHeight('50px')
      .addTextBlock((b) => b.withText('Hello'))
      .addImage((b) => b.withUrl('https://example.com/img.png'))
      .build();

    assert.equal(container.type, 'Container');
    assert.equal(container.id, 'c1');
    assert.equal(container.style, ContainerStyle.Emphasis);
    assert.equal(container.verticalContentAlignment, VerticalAlignment.Center);
    assert.equal(container.bleed, true);
    assert.equal(container.minHeight, '50px');
    assert.equal(container.items!.length, 2);
  });

  it('addElement accepts a pre-built element', () => {
    const img = new (require('fluent-cards').ImageBuilder)().withUrl('https://example.com/x.png').build();
    const container = new ContainerBuilder().addElement(img).build();
    assert.equal(container.items!.length, 1);
    assert.equal(container.items![0].type, 'Image');
  });

  it('is added to card body via addContainer', () => {
    const card = AdaptiveCardBuilder.create()
      .addContainer((c) => c.addTextBlock((b) => b.withText('Nested')))
      .build();

    assert.equal(card.body![0].type, 'Container');
    assert.equal(((card.body![0] as Container).items![0] as any).text, 'Nested');
  });

  it('serializes container style as camelCase', () => {
    const card = AdaptiveCardBuilder.create()
      .addContainer((c) => c.withStyle(ContainerStyle.Accent).addTextBlock((b) => b.withText('x')))
      .build();
    assert.ok(toJson(card).includes('"style": "accent"'));
  });
});

describe('ColumnSetBuilder', () => {
  it('builds columns with the two-argument overload', () => {
    const cs = new ColumnSetBuilder()
      .addColumn('auto', (col) => col.addTextBlock((b) => b.withText('A')))
      .addColumn('stretch', (col) => col.addTextBlock((b) => b.withText('B')))
      .build();

    assert.equal(cs.type, 'ColumnSet');
    assert.equal(cs.columns!.length, 2);
    assert.equal(cs.columns![0].width, 'auto');
    assert.equal(cs.columns![1].width, 'stretch');
  });

  it('builds columns with the single-argument overload', () => {
    const cs = new ColumnSetBuilder()
      .addColumn((col) => col.withWidth('1').addTextBlock((b) => b.withText('Col')))
      .build();

    assert.equal(cs.columns!.length, 1);
    assert.equal(cs.columns![0].width, '1');
  });

  it('serializes columns', () => {
    const card = AdaptiveCardBuilder.create()
      .addColumnSet((cs) =>
        cs
          .addColumn('auto', (col) => col.addTextBlock((b) => b.withText('Left')))
          .addColumn('auto', (col) => col.addTextBlock((b) => b.withText('Right'))),
      )
      .build();

    const json = toJson(card);
    assert.ok(json.includes('"type": "ColumnSet"'));
    assert.ok(json.includes('"type": "Column"'));
    assert.ok(json.includes('"text": "Left"'));
    assert.ok(json.includes('"text": "Right"'));
  });
});

describe('FactSetBuilder', () => {
  it('builds FactSet with string overload', () => {
    const fs = new FactSetBuilder()
      .addFact('Name', 'Alice')
      .addFact('Age', '30')
      .build();

    assert.equal(fs.type, 'FactSet');
    assert.equal(fs.facts!.length, 2);
    assert.equal(fs.facts![0].title, 'Name');
    assert.equal(fs.facts![0].value, 'Alice');
  });

  it('builds FactSet with object overload', () => {
    const fs = new FactSetBuilder()
      .addFact({ title: 'City', value: 'NYC' })
      .build();

    assert.equal(fs.facts![0].title, 'City');
  });
});

describe('RichTextBlockBuilder', () => {
  it('builds inlines with strings and TextRuns', () => {
    const rtb = new RichTextBlockBuilder()
      .addText('Plain text ')
      .addTextRun((r) => r.withText('Bold').withWeight(TextWeight.Bolder).withColor(TextColor.Accent))
      .build();

    assert.equal(rtb.type, 'RichTextBlock');
    assert.equal(rtb.inlines!.length, 2);
    assert.equal(rtb.inlines![0], 'Plain text ');
    const run = rtb.inlines![1] as any;
    assert.equal(run.type, 'TextRun');
    assert.equal(run.text, 'Bold');
    assert.equal(run.weight, TextWeight.Bolder);
    assert.equal(run.color, TextColor.Accent);
  });

  it('serializes RichTextBlock inlines', () => {
    const card = AdaptiveCardBuilder.create()
      .addRichTextBlock((b) =>
        b
          .addText('Hello ')
          .addTextRun((r) => r.withText('World').withWeight(TextWeight.Bolder)),
      )
      .build();
    const json = toJson(card);
    assert.ok(json.includes('"type": "RichTextBlock"'));
    assert.ok(json.includes('"type": "TextRun"'));
    assert.ok(json.includes('"Hello "'));
    assert.ok(json.includes('"weight": "bolder"'));
  });
});

describe('MediaBuilder', () => {
  it('builds a Media element', () => {
    const media = new MediaBuilder()
      .withId('m1')
      .withPoster('https://example.com/poster.jpg')
      .withAltText('A video')
      .addSource('https://example.com/video.mp4', 'video/mp4')
      .addSource('https://example.com/video.webm', 'video/webm')
      .build();

    assert.equal(media.type, 'Media');
    assert.equal(media.poster, 'https://example.com/poster.jpg');
    assert.equal(media.sources!.length, 2);
    assert.equal(media.sources![0].mimeType, 'video/mp4');
  });
});

describe('ImageSetBuilder', () => {
  it('builds an ImageSet with images', () => {
    const is = new ImageSetBuilder()
      .withImageSize(ImageSize.Medium)
      .addImage((b) => b.withUrl('https://example.com/a.png'))
      .addImage((b) => b.withUrl('https://example.com/b.png'))
      .build();

    assert.equal(is.type, 'ImageSet');
    assert.equal(is.imageSize, ImageSize.Medium);
    assert.equal(is.images!.length, 2);
  });
});

describe('TableBuilder', () => {
  it('builds a Table with columns and rows', () => {
    const col: TableColumnDefinition = { width: '1' };
    const cell: TableCell = { type: 'TableCell', items: [{ type: 'TextBlock', text: 'Cell' }] };
    const row: TableRow = { type: 'TableRow', cells: [cell] };

    const table = new TableBuilder()
      .withFirstRowAsHeader()
      .withShowGridLines()
      .addColumn(col)
      .addRow(row)
      .build();

    assert.equal(table.type, 'Table');
    assert.equal(table.firstRowAsHeader, true);
    assert.equal(table.showGridLines, true);
    assert.equal(table.columns!.length, 1);
    assert.equal(table.rows!.length, 1);
  });

  it('serializes Table correctly', () => {
    const card = AdaptiveCardBuilder.create()
      .addTable((t) => {
        t.withFirstRowAsHeader();
        t.addColumn({ width: 'auto' });
        t.addRow({
          type: 'TableRow',
          cells: [{ type: 'TableCell', items: [{ type: 'TextBlock', text: 'Header' }] }],
        });
      })
      .build();

    const json = toJson(card);
    assert.ok(json.includes('"type": "Table"'));
    assert.ok(json.includes('"type": "TableRow"'));
    assert.ok(json.includes('"type": "TableCell"'));
    assert.ok(json.includes('"firstRowAsHeader": true'));
  });
});

describe('ActionSetBuilder', () => {
  it('builds an ActionSet with actions', () => {
    const as = new ActionSetBuilder()
      .withId('as1')
      .addAction((b) => b.openUrl('https://example.com', 'Visit'))
      .build();

    assert.equal(as.type, 'ActionSet');
    assert.equal(as.id, 'as1');
    assert.equal(as.actions!.length, 1);
    assert.equal(as.actions![0].type, 'Action.OpenUrl');
  });
});

File: go/fluentcards/elements_test.go
package fluentcards_test

import (
	"testing"

	"github.com/rido-min/FluentCards/go/fluentcards"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestTextBlockBuilder_AllProperties(t *testing.T) {
	t.Parallel()
	card := fluentcards.NewAdaptiveCardBuilder().
		AddTextBlock(func(tb *fluentcards.TextBlockBuilder) {
			tb.WithText("Hello").
				WithSize(fluentcards.TextSizeLarge).
				WithWeight(fluentcards.TextWeightBolder).
				WithColor(fluentcards.TextColorAccent).
				WithWrap(true).
				WithMaxLines(3).
				WithIsSubtle(false).
				WithHorizontalAlignment(fluentcards.HorizontalAlignmentCenter).
				WithFontType(fluentcards.FontTypeMonospace).
				WithStyle(fluentcards.TextBlockStyleHeading).
				WithSpacing(fluentcards.SpacingMedium).
				WithSeparator(true).
				WithID("tb1")
		}).Build()
	el := card["body"].([]any)[0].(map[string]any)
	assert.Equal(t, "TextBlock", el["type"])
	assert.Equal(t, "Hello", el["text"])
	assert.Equal(t, "large", el["size"])
	assert.Equal(t, "bolder", el["weight"])
	assert.Equal(t, "accent", el["color"])
	assert.Equal(t, true, el["wrap"])
	assert.Equal(t, 3, el["maxLines"])
	assert.Equal(t, "center", el["horizontalAlignment"])
	assert.Equal(t, "monospace", el["fontType"])
	assert.Equal(t, "heading", el["style"])
	assert.Equal(t, "medium", el["spacing"])
	assert.Equal(t, true, el["separator"])
	assert.Equal(t, "tb1", el["id"])
}

func TestImageBuilder_Properties(t *testing.T) {
	t.Parallel()
	card := fluentcards.NewAdaptiveCardBuilder().
		AddImage(func(img *fluentcards.ImageBuilder) {
			img.WithURL("https://example.com/img.png").
				WithAltText("An image").
				WithSize(fluentcards.ImageSizeMedium).
				WithStyle(fluentcards.ImageStylePerson).
				WithWidth("100px").
				WithHeight("100px").
				WithBackgroundColor("#FFFFFF")
		}).Build()
	el := card["body"].([]any)[0].(map[string]any)
	assert.Equal(t, "Image", el["type"])
	assert.Equal(t, "https://example.com/img.png", el["url"])
	assert.Equal(t, "An image", el["altText"])
	assert.Equal(t, "medium", el["size"])
	assert.Equal(t, "person", el["style"])
	assert.Equal(t, "100px", el["width"])
	assert.Equal(t, "#FFFFFF", el["backgroundColor"])
}

func TestContainerBuilder_WithItems(t *testing.T) {
	t.Parallel()
	card := fluentcards.NewAdaptiveCardBuilder().
		AddContainer(func(c *fluentcards.ContainerBuilder) {
			c.WithStyle(fluentcards.ContainerStyleEmphasis).
				WithBleed(true).
				AddTextBlock(func(tb *fluentcards.TextBlockBuilder) { tb.WithText("Inside container") }).
				AddImage(func(img *fluentcards.ImageBuilder) { img.WithURL("https://example.com/x.png") })
		}).Build()
	el := card["body"].([]any)[0].(map[string]any)
	assert.Equal(t, "Container", el["type"])
	assert.Equal(t, "emphasis", el["style"])
	assert.Equal(t, true, el["bleed"])
	items := el["items"].([]any)
	assert.Len(t, items, 2)
}

func TestColumnSetBuilder_WithColumns(t *testing.T) {
	t.Parallel()
	card := fluentcards.NewAdaptiveCardBuilder().
		AddColumnSet(func(cs *fluentcards.ColumnSetBuilder) {
			cs.AddColumnWithWidth("auto", func(col *fluentcards.ColumnBuilder) {
				col.AddTextBlock(func(tb *fluentcards.TextBlockBuilder) { tb.WithText("Left") })
			}).AddColumnWithWidth("stretch", func(col *fluentcards.ColumnBuilder) {
				col.WithVerticalContentAlignment(fluentcards.VerticalAlignmentCenter).
					AddTextBlock(func(tb *fluentcards.TextBlockBuilder) { tb.WithText("Right") })
			})
		}).Build()
	el := card["body"].([]any)[0].(map[string]any)
	assert.Equal(t, "ColumnSet", el["type"])
	cols := el["columns"].([]any)
	require.Len(t, cols, 2)
	assert.Equal(t, "auto", cols[0].(map[string]any)["width"])
	assert.Equal(t, "stretch", cols[1].(map[string]any)["width"])
}

func TestFactSetBuilder_AddFact(t *testing.T) {
	t.Parallel()
	card := fluentcards.NewAdaptiveCardBuilder().
		AddFactSet(func(fs *fluentcards.FactSetBuilder) {
			fs.AddFact("Name", "Alice").
				AddFact("Role", "Engineer")
		}).Build()
	el := card["body"].([]any)[0].(map[string]any)
	facts := el["facts"].([]any)
	require.Len(t, facts, 2)
	assert.Equal(t, "Name", facts[0].(map[string]any)["title"])
	assert.Equal(t, "Alice", facts[0].(map[string]any)["value"])
}

func TestRichTextBlockBuilder_Inlines(t *testing.T) {
	t.Parallel()
	card := fluentcards.NewAdaptiveCardBuilder().
		AddRichTextBlock(func(rtb *fluentcards.RichTextBlockBuilder) {
			rtb.AddText("plain text").
				AddTextRun(func(tr *fluentcards.TextRunBuilder) {
					tr.WithText("bold").WithWeight(fluentcards.TextWeightBolder).WithItalic(true)
				})
		}).Build()
	el := card["body"].([]any)[0].(map[string]any)
	assert.Equal(t, "RichTextBlock", el["type"])
	inlines := el["inlines"].([]any)
	require.Len(t, inlines, 2)
	assert.Equal(t, "plain text", inlines[0])
	run := inlines[1].(map[string]any)
	assert.Equal(t, "TextRun", run["type"])
	assert.Equal(t, "bold", run["text"])
	assert.Equal(t, "bolder", run["weight"])
	assert.Equal(t, true, run["italic"])
}

func TestActionSetBuilder_WithActions(t *testing.T) {
	t.Parallel()
	card := fluentcards.NewAdaptiveCardBuilder().
		AddActionSet(func(as *fluentcards.ActionSetBuilder) {
			as.AddAction(func(a *fluentcards.ActionBuilder) { a.Submit("OK") }).
				AddAction(func(a *fluentcards.ActionBuilder) { a.OpenURL("https://example.com") })
		}).Build()
	el := card["body"].([]any)[0].(map[string]any)
	assert.Equal(t, "ActionSet", el["type"])
	actions := el["actions"].([]any)
	assert.Len(t, actions, 2)
}

func TestMediaBuilder_WithSources(t *testing.T) {
	t.Parallel()
	card := fluentcards.NewAdaptiveCardBuilder().
		AddMedia(func(m *fluentcards.MediaBuilder) {
			m.WithPoster("https://example.com/poster.png").
				AddSource("https://example.com/video.mp4", "video/mp4")
		}).Build()
	el := card["body"].([]any)[0].(map[string]any)
	assert.Equal(t, "Media", el["type"])
	sources := el["sources"].([]any)
	require.Len(t, sources, 1)
	s := sources[0].(map[string]any)
	assert.Equal(t, "https://example.com/video.mp4", s["url"])
	assert.Equal(t, "video/mp4", s["mimeType"])
}

func TestImageSetBuilder_AddImages(t *testing.T) {
	t.Parallel()
	card := fluentcards.NewAdaptiveCardBuilder().
		AddImageSet(func(is *fluentcards.ImageSetBuilder) {
			is.WithImageSize(fluentcards.ImageSizeMedium).
				AddImage(func(img *fluentcards.ImageBuilder) { img.WithURL("https://example.com/1.png") }).
				AddImage(func(img *fluentcards.ImageBuilder) { img.WithURL("https://example.com/2.png") })
		}).Build()
	el := card["body"].([]any)[0].(map[string]any)
	assert.Equal(t, "ImageSet", el["type"])
	assert.Equal(t, "medium", el["imageSize"])
	images := el["images"].([]any)
	assert.Len(t, images, 2)
}

func TestTableBuilder_AddColumnsAndRows(t *testing.T) {
	t.Parallel()
	card := fluentcards.NewAdaptiveCardBuilder().
		AddTable(func(tb *fluentcards.TableBuilder) {
			tb.WithFirstRowAsHeader(true).
				WithShowGridLines(true).
				AddColumn(map[string]any{"width": 1}).
				AddColumn(map[string]any{"width": 2}).
				AddRow(map[string]any{
					"cells": []any{
						map[string]any{"items": []any{map[string]any{"type": "TextBlock", "text": "H1"}}},
						map[string]any{"items": []any{map[string]any{"type": "TextBlock", "text": "H2"}}},
					},
				})
		}).Build()
	el := card["body"].([]any)[0].(map[string]any)
	assert.Equal(t, "Table", el["type"])
	assert.Equal(t, true, el["firstRowAsHeader"])
	cols := el["columns"].([]any)
	assert.Len(t, cols, 2)
	rows := el["rows"].([]any)
	assert.Len(t, rows, 1)
}
